### PR TITLE
First addition of butler requirements

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

**Comment by r-owen** [comment]
Comments on items through DMS-MWBT-REQ-0055.

These are mostly in addition to comments posted on dm-butler-wg (though there may be a few duplications).

Terms such as DataUnit and ConcreteDataset are not defined here and are not presently used in the project. Even terms such as Data Input System may not be well understood by all readers. I strongly suggest a section of the document that describes the technical terms used in the requirements, with links to other documents for more information.

DMS-MWBT-REQ-0001 1.1.9 discussion says “Removing the repository might require database table modifications.” I don’t think it’s relevant here. I suggest moving it to the discussion of the requirement on Repository Removal or ditching it.

In DMS-MWBT-REQ-0003 1.1.12 Repository version migration what is meant by “perform *persistent migrations”? I think the text should be clarified and I hope “persistent” can be left out. Consider: “The Data Input/Output system shall be able to migrate a DataRepository…“.

DMS-MWBT-REQ-0010 Subsetting a DataRepository without data transfer: do we have a use case for this? The two use cases that are referenced both explicitly copy data (which is covered by the previous requirement). I think we need to rewrite one or both use cases a bit, add a new use case, or ditch the requirement.

DMS-MWBT-REQ-0017 Data Repository Layering: Science Platform “Data Repositories created in the Science Platform (A) shall be usable as inputs for processing initiated in the Science Platform (B).” What’s with the (A) and (B)? I hope they can be removed, but if not, please add some clarifying text.

Likewise for the next requirement: DMS-MWBT-REQ-0018

DMS-MWBT-REQ-0090 Filter by config “...based upon user-specified filters provided via text file containing...”. Just add an “a” before “text file”. Also please consider not requiring that it be a text file. What matters is that there is some way for a user to do this, with the filter saved somewhere so it can be used again. A text file is certainly a logical choice and could be mentioned in the discussion.

DMS-MWBT-REQ-0082 Multiple parallel input DataRepositories “I don’t believe this is a requirement on whether this is two butlers or one, but I do believe some other requirements imply a single system that knows about all repositories in a particular context.”. This is clear, but I suggest rewording without “I” since these requirements do not have an identified single author.

DMS-MWBT-REQ-0026 typo: A mechanism shall be availble
DMS-MWBT-REQ-0025 Format pluggability “It shall be possible to control the method to use for read and write of a particular DatasetType using a text configuration file such that the Python object and the form of the persisted dataset can be configured externally.” This could use editing. At least consider “It shall be possible to control the method used to read and write a particular DatasetType...”. I suspect two sentences would be better, but didn’t see an easy way to break it up.
DMS-MWBT-REQ-0031 is “cloudy storage engines” really a thing? I was expecting “cloud storage” e.g. “...utilize cloud storage engines such as S3”.
DMS-MWBT-REQ-0056 “...different Tasks may refer to the same table-like dataset by name”. This is about naming *subsets* of catalogs (we don’t need a requirement for referring to an entire catalog). Thus I think the wording should be changed a bit. Here is one option: “...same subset of a catalog by name”. (I think “...same subset of a table-like dataset by name” is too verbose.)
Also the discussion uses “I” and so should be reworded.
Also the discussion says “It would be nice to support parameterized views”. Please add an explanation. I’m guessing something like “FWHM < x” where the user can somehow specify “x”.
DMS-MWBT-REQ-0053 “For example, the batch processing system will have retrieved a valid set of files from the Data Backbone and copied them to a local disk”. This is picky, but I suggest rewording and additional clarification: “...will retrieve a valid set of files from...and copy them to a local disk. The Data Input system will then read them from local disk.”
DMS-MWBT-REQ-0046 and DMS-MWBT-REQ-0047 could use a real discussion instead of just “(UseCases:...)”.
DMS-MWBT-REQ-0054 Failure missing input: 
The specification should state which system or systems this requirement applies to. The Data Input System? Any other systems?
I find this a bit confusing. Do we consider it a requirement that this system can be configured to do anything other than fail if a file is missing (e.g. search the data backbone, as implied by the discussion)? Or are we merely hedging our bets: e.g. if the system can try to find missing data then we must be able to disable that feature?
At least please tweak the title E.g. “Fail on missing input” and text, e.g. “to require failure if file not found at specified location” -> “to configure the (Data Input System?) to fail a file is not found at a specified location”.

DMS-MWBT-REQ-0052 Input Staging: I am puzzled by the discussion. If the Operations Batch Processing Service isn’t going to use this, then who is? It would help to say in the discussion what systems will use this!
Note that DRP5, DRP6 are listed as relevant use cases, but they are bout the Batch Control System, which I assume is another name for the Operations Batch Processing Service? If so, then the discussion negates says they are not relevant. The SCIVAL may still drive this use case; I confess I’m not sure.
If Batch Control System and Operations Batch Processing Service are the same thing then can we please pick one term and stick to it for both the use cases and requirements?
DMS-MWBT-REQ-0058 Local caching of remote resources. Is this really 1a? It sounds more like a “nice to have” than a requirement, since we can copy a subset of a data repo.
DMS-MWBT-REQ-0055 Local Proxy: please explain further what it means to "configure the Data Input system to use a local proxy”. The discussion says “Thus allowing multiple local input requests (potentially from different users and systems) to use a shared cache.” but I don’t understand what is being cached. If possible, I suggest defining “local proxy” in the Specification section, since that definition sounds like an important part of the requirement.

**Comment by timj** [comment]
Thanks for the comments and I have made some updates to the model. I'm again concerned that comments as substantial as these arrive long after we thought the requirements were settled.

Some comments and requests for feedback:

* For 0003 I'm assuming "persistent" is there to imply that you can't easily migrate back to an older version of the repository.
* We talked about a new Use Case for 0010. The question is who is going to write it. Do we favor a CI use case where the CI job is always using a subset of the whole (so you can have it include specific CCDs from the full raw archive for example without having to fully-specify everything).
* @pschella @TallJimbo the (A) and (B) stuff in the Data Repository Layering requirements is a bit confusing.
* @SimonKrughoff in 0056 are the parameterized views the ability to adjust the view from within a butler configuration file rather than explicitly defining the view in an external context and then using the name in a butler config file?
* @TallJimbo @pschella do you feel 0046 and 0047 need additional discussion? (without copying DRP11 into the requirement). Maybe an example non-LSST dataset?
* Re 0054, I think the point is that in some cases the butler might go looking for a file if it can't be found. @MichelleGower wanted to make it absolutely clear that it should be possible to configure a butler so that it won't do that.
* 0052 The feature is clearly useful, it's just that @MichelleGower did not want to commit to using it at the LDF. I can imagine it being useful for developers regardless. We clearly need to be able to take the preflight request from the SuperTask and work out what files to transfer. Maybe the "Note" should be dropped though as it doesn't really help anything and since there are no LDF UseCases that's implied. It is true that DRP5, DRP6, and SCIVAL1 look suspiciously like they are talking about the batch processing system in the LDF so there is a mismatch there. @TallJimbo maybe you have a comment here on this requirement outside the context of LDF using it for a DRP run.

**Comment by TallJimbo** [comment]
Replying to @timj:

> For 0003 I'm assuming "persistent" is there to imply that you can't easily migrate back to an older version of the repository.

That is my understanding.

> We talked about a new Use Case for 0010. The question is who is going to write it. Do we favor a CI use case where the CI job is always using a subset of the whole (so you can have it include specific CCDs from the full raw archive for example without having to fully-specify everything).

Do the existing use cases not need this requirement or something?  `DRP16` does, which might remove the need for a new one (and should probably be noted anyway).

**Comment by TallJimbo** [comment]
More replies to @timj (hit enter too early):

> the (A) and (B) stuff in the Data Repository Layering requirements is a bit confusing.

Since you've just duplicated the text, I think we can just drop all of the (A) and (B) labels without losing anything.

> do you feel 0046 and 0047 need additional discussion? (without copying DRP11 into the requirement). Maybe an example non-LSST dataset?

Could we just copy the last three sentences of DRP11?  Those are pretty comprehensive and already pretty requirements-ey.

> 0052 The feature is clearly useful, it's just that @MichelleGower did not want to commit to using it at the LDF.

I'm okay with just dropping the requirement.  The design will support that functionality anyway, since essentially the same logic is demanded by other transfer requirements (e.g. 0010), and I'm happy to defer to @MichelleGower as to whether we actually need it (or, rather, whether we consider it part of the Data Input System or not).

**Comment by r-owen** [comment]
More comments on the document. At this point I've been through all the use cases, so expect to be done with big blocks of feedback.

DMS-MWBT-REQ-0048 “It shall be possible for the Data Input System to read from catalogs provided by outside sources in the same API used for reading first class LSST datasets...”. I suggest a bit of rewording, as I’m not sure what “in the same API” means.

DMS-MWBT-REQ-0038 and 0039: both of these are about output, but they specify a requirement on the Data INPUT System Interface. A bit of clarification of the text of both (and possibly also their titles) would be very helpful. Are these intended to be a requirement that the Data Output System can write the output from notebook batch jobs and test processing runs? If so, the title and text should say something like “Access to outputs from...”. Or if this is about the ability of our code to write the outputs (as implied by the current titles), then perhaps it is the Data OUTPUT System that should be mentioned?

Also, as a smaller point, does the word “Interface” in “Data Input System Interface” provide important information? If not, I suggest deleting it.

DMS-MWBT-REQ-0074 "such that it is an error if an attempt is made to persist a dataset” could be tightened up a bit, e.g.: “such that it is an error to attempt to persist a dataset”.

DMS-MWBT-REQ-0066 “It shall be possible to configure the Data Ouput System via configuration to define”:remove the redundant “via configuration” and add the a missing “t” in “Ouput”.

DMS-MWBT-REQ-0069 “...being run as part of a Alert Production”: delete the “a” before “Alert”

DMS-MWBT-REQ-0068 “...“a Data Release Production”: again, delete the “a” before “Data”.

DMS-MWBT-REQ-0070 “code run by a batch serves in the Science Platform”: should it be “a batch service”? (If not, please reword “a batch serves” in some other way).

DMS-MWBT-REQ-0095 in the discussion, please provide more words to explain this concluding fragment: ”Trickle up provenance” for ”was derived from”.

Task framework starts with “The following requirements have been developed by the SuperTask working group as design requirements that derive from the requirements identified in the working group.” Please reword to remove redundancy and please clarify which working group is meant at the end. (If it is the SuperTask working group then I hope the final mention can be deleted).

DMS-MWST-REQ-0021 please spell out DAG the first time it is used.

DMS-MWST-REQ-0029 “I arbitrarily allocated” -> “We arbitrarily allocated”.

3.1.3 “The LDF shall make it possible, given sufficient permissions, do output to a persistent...”. “to output” instead of “do output”? I suggest rewriting this as at least two sentences.
